rfcs/text/0015_nodejs_clustering.md

```diff
@@ -0,0 +1,641 @@
+### 6.2.2 Identified required changes
+
+#### Reporting
+
+We will probably want to restrict to a single headless per Kibana instance. For that, we will have to change the logic 
+in [createQueueFactory](https://github.com/elastic/kibana/blob/4584a8b570402aa07832cf3e5b520e5d2cfa7166/x-pack/plugins/reporting/server/lib/create_queue.ts#L60-L64) 
+to only have the 'main' worker be pooling for reporting tasks.
+
+#### Telemetry
+
+- Server side fetcher
+  
+The telemetry/server/fetcher.ts will attempt sending the telemetry usage multiple times once per day from each process. 
+We do store a state in the SavedObjects store of the last time the usage was sent to prevent sending multiple times 
+(although race conditions might occur).
+
+- Tasks storing telemetry data
+  
+We have tasks across several plugins storing data in savedobjects specifically for telemetry. Under clustering these 
+tasks will be registered multiple times.
+
+Note that sending the data multiple times doesn’t have any real consequences, apart from the additional number of ES requests, 
+so this should be considered non-blocking and only an improvement.
+
+- Event-based telemetry
+
+Event-based telemetry may be affected as well. Both the existing one in the Security Solutions team and the general 
+one that is in the works. More specifically, the size of the queues will be multiplied per worker, also growing in the
+amount of network bandwidth used, and potentially affecting our customers.
+
+We could address that by making sure that the queues are held only in the main worker.
+
+#### Task Manager
+
+Currently, task manager does "claims" for jobs to run based on the server uuid. We think this could still work with 
+a multi-process setup - each task manager in the worker would be doing "claims" for the same server uuid, which
+seems functionally the same as setting max_workers to `current max_workers * number of workers`.
+
+However, as a first step we can simply run Task Manager on the main worker. This doesn't completely solve potential
+noisy neighbor problems as the main worker will still be receiving & serving http requests, however it will at least
+ensure that other worker processes are free to serve http requests without risk of TM interference. Long term, we
+could explore manually spawning a dedicated child process for background tasks that can be called from workers, and
+thinking of a way for plugins to tell Core when they need to run things in the background.
+
+It would be ideal if we could eventually solve this with our multi-process setup, however this needs more design work
+and could necessitate an RFC in its own right. The key thing to take away here is that the work we are doing in this
+RFC would not prevent us from exploring this path further in a subsequent phase. In fact, it could prove to be a
+helpful first step in that direction.
+
+#### Alerting
+
+Do we need the alerting task runner (`TaskRunner`, `TaskRunnerFactory`) to be running on a single worker?
```

Comment:
Not that I can recall, it would inherit the same requirements as the task manager section.

Reply:
I might have misread the question, is this question about having workers that only run alerting tasks? (nothing else)

Reply:
@mikecote @gmmorris I think the question should be generalized wider than just alerting tasks. If a task is defined with a `maxConcurrency` option (which only Reporting at this moment is AFAIK), is the number of running tasks allowed still controlled at the Kibana instance level, or the worker level? 

I don't have an opinion either way. I think it will be OK for Reporting if the number of running tasks is limited by the number of cores.

Reply:
Ah, this is a good point.
It's worker level, if I understand the terminology correctly.
Basically, if each core is running its own Task Manager, then there could easily be multiple reporting tasks running on the same machine (up to 1 per core).